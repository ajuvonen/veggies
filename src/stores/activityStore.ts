import {computed, ref} from 'vue';
import {defineStore, storeToRefs} from 'pinia';
import {debounceFilter, useIntervalFn, useNow, useStorage} from '@vueuse/core';
import {DateTime} from 'luxon';
import {
  countBy,
  difference,
  entries,
  intersection,
  map,
  pipe,
  prop,
  sortBy,
  take,
  takeWhile,
} from 'remeda';
import {
  Category,
  type Favorites,
  type Challenge,
  type Week,
  type Achievements,
  AchievementLevel,
} from '@/utils/types';
import {
  achievementLevelHelper,
  dateParser,
  dateReplacer,
  getCategoryForVeggie,
  getRandomVeggie,
} from '@/utils/helpers';
import {useAppStateStore} from '@/stores/appStateStore';
import {
  BEANS,
  BOTANICAL_BERRIES,
  FRUITS,
  GRAINS,
  LEAFIES,
  MUSHROOMS,
  NUTS,
  RED_VEGGIES,
  ROOTS,
  VEGETABLES,
} from '@/utils/constants';

export const useActivityStore = defineStore('activity', () => {
  const {settings} = storeToRefs(useAppStateStore());
  const currentDate = ref(DateTime.now());
  useIntervalFn(() => {
    const now = DateTime.now();
    if (!currentDate.value.hasSame(now, 'day')) {
      currentDate.value = now;
    }
  }, 2000);

  // State refs
  const startDate = useStorage<DateTime | null>('veggies-start-date', null, localStorage, {
    mergeDefaults: true,
    serializer: {
      read: (v) => (v ? DateTime.fromISO(v.split('T')[0]) : null),
      write: (v) => v?.toISODate() ?? '',
    },
  });

  const weeks = useStorage<Week[]>('veggies-weeks', [], localStorage, {
    mergeDefaults: true,
    eventFilter: debounceFilter(2000),
    serializer: {
      read: (v) => (v ? JSON.parse(v, dateParser) : null),
      write: (v) => JSON.stringify(v, dateReplacer),
    },
  });

  const challenges = useStorage<Challenge[]>('veggies-challenges', [], localStorage, {
    mergeDefaults: true,
    serializer: {
      read: (v) => (v ? JSON.parse(v, dateParser) : null),
      write: (v) => JSON.stringify(v, dateReplacer),
    },
  });

  // Computed getters
  const currentWeekStart = computed(() => currentDate.value.startOf('week'));

  const getWeekStarts = computed(() => {
    if (!startDate.value) return [currentWeekStart.value];
    const totalWeeks = Math.ceil(currentDate.value.diff(startDate.value, 'week').weeks);
    return Array.from({length: totalWeeks}, (_, weekIndex) =>
      currentWeekStart.value.minus({weeks: weekIndex}),
    );
  });

  const hotStreak = computed(() => {
    let over30Weeks = takeWhile(
      getWeekStarts.value,
      (weekStart, weekIndex) => weekIndex === 0 || veggiesForWeek.value(weekStart).length >= 30,
    ).length;
    if (currentVeggies.value.length < 30) {
      over30Weeks--;
    }
    return over30Weeks;
  });

  const allVeggies = computed(() => weeks.value.flatMap(prop('veggies')));

  const uniqueVeggies = computed(() => [...new Set(allVeggies.value)]);

  const over30Veggies = computed(
    () => weeks.value.filter(({veggies}) => veggies.length >= 30).length,
  );

  const atMostVeggies = computed(() =>
    Math.max(...weeks.value.map(({veggies}) => veggies.length), 0),
  );

  const completedChallenges = computed(
    () =>
      challenges.value.filter(({startDate, veggie}) =>
        veggiesForWeek.value(startDate).includes(veggie),
      ).length,
  );

  const veggiesForWeek = computed(
    () => (weekStart: DateTime) =>
      weeks.value.find(({startDate}) => startDate.equals(weekStart))?.veggies ?? [],
  );

  const currentVeggies = computed({
    get: () => veggiesForWeek.value(currentWeekStart.value),
    set: (veggies: string[]) => setVeggiesForWeek(veggies),
  });

  const currentChallenge = computed(
    () => challenges.value.find(({startDate}) => startDate.equals(currentWeekStart.value))?.veggie,
  );

  const suggestions = computed(() =>
    pipe(
      allVeggies.value.filter((veggie) => !currentVeggies.value.includes(veggie)),
      countBy((veggie) => veggie),
      entries(),
      sortBy([prop(1), 'desc']),
      take(settings.value.suggestionCount),
      map(prop(0)),
    ),
  );

  const favorites = computed(() =>
    Object.values(Category).reduce(
      (acc, category) => ({
        ...acc,
        [category]: pipe(
          allVeggies.value.filter((veggie) => getCategoryForVeggie(veggie) === category),
          countBy((veggie) => veggie),
          entries(),
          sortBy([prop(1), 'desc']),
          take(6),
        ),
      }),
      {} as Favorites,
    ),
  );

  const weeklyAchievements = computed(() => (veggies: string[] = currentVeggies.value) => ({
    allOnRed: achievementLevelHelper(
      [[10, AchievementLevel.Gold]],
      intersection(RED_VEGGIES, veggies).length,
    ),
    botanicalBerries: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(BOTANICAL_BERRIES, veggies).length,
    ),
    goNuts: achievementLevelHelper(
      [[5, AchievementLevel.Gold]],
      intersection(NUTS, veggies).length,
    ),
    thirtyVeggies: achievementLevelHelper(
      [
        [40, AchievementLevel.Platinum],
        [30, AchievementLevel.Gold],
      ],
      veggies.length,
    ),
  }));

  const achievements = computed<Achievements>(() => ({
    challengeAccepted: achievementLevelHelper(
      [
        [20, AchievementLevel.Gold],
        [10, AchievementLevel.Silver],
        [5, AchievementLevel.Bronze],
      ],
      completedChallenges.value,
    ),
    committed: achievementLevelHelper(
      [
        [52, AchievementLevel.Gold],
        [26, AchievementLevel.Silver],
        [12, AchievementLevel.Bronze],
      ],
      weeks.value.length,
    ),
    completionist: achievementLevelHelper(
      [
        [150, AchievementLevel.Gold],
        [80, AchievementLevel.Silver],
        [40, AchievementLevel.Bronze],
      ],
      uniqueVeggies.value.length,
    ),
    experimenterBean: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(BEANS, uniqueVeggies.value).length,
    ),
    experimenterFruit: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(FRUITS, uniqueVeggies.value).length,
    ),
    experimenterGrain: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(GRAINS, uniqueVeggies.value).length,
    ),
    experimenterLeafy: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(LEAFIES, uniqueVeggies.value).length,
    ),
    experimenterMushroom: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(MUSHROOMS, uniqueVeggies.value).length,
    ),
    experimenterRoot: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(ROOTS, uniqueVeggies.value).length,
    ),
    experimenterVegetable: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(VEGETABLES, uniqueVeggies.value).length,
    ),
    hotStreak: achievementLevelHelper(
      [
        [20, AchievementLevel.Gold],
        [10, AchievementLevel.Silver],
        [5, AchievementLevel.Bronze],
      ],
      hotStreak.value,
    ),
    thousandsEven:
      allVeggies.value.length >= 1000 && Math.floor(allVeggies.value.length / 1000) % 2 === 0
        ? AchievementLevel.Platinum
        : AchievementLevel.NoAchievement,
    thousandsOdd:
      allVeggies.value.length >= 1000 && Math.floor(allVeggies.value.length / 1000) % 2 === 1
        ? AchievementLevel.Platinum
        : AchievementLevel.NoAchievement,
    ...weeklyAchievements.value(),
  }));

  // Actions
  const toggleVeggie = (targetVeggie: string) =>
    toggleVeggieForWeek(targetVeggie, currentWeekStart.value);

  const toggleVeggieForWeek = (targetVeggie: string, weekStart: DateTime) => {
    const targetWeek = weeks.value.find(({startDate}) => startDate.equals(weekStart));
    if (!targetWeek) {
      weeks.value = [
        ...weeks.value,
        {
          startDate: weekStart,
          veggies: [targetVeggie],
        },
      ];
      challenges.value = [
        ...challenges.value,
        {
          startDate: weekStart,
          veggie: getRandomVeggie(),
        },
      ];
    } else if (!targetWeek.veggies.includes(targetVeggie)) {
      targetWeek.veggies = [...targetWeek.veggies, targetVeggie];
    } else {
      targetWeek.veggies = difference(targetWeek.veggies, [targetVeggie]);
    }
  };

  const setVeggiesForWeek = (veggies: string[], weekStart: DateTime = currentWeekStart.value) => {
    const targetWeek = weeks.value.find(({startDate}) => startDate.equals(weekStart));
    if (!targetWeek) {
      weeks.value = [
        ...weeks.value,
        {
          startDate: weekStart,
          veggies,
        },
      ];
      challenges.value = [
        ...challenges.value,
        {
          startDate: weekStart,
          veggie: getRandomVeggie(),
        },
      ];
    } else {
      targetWeek.veggies = veggies;
    }
  };

  const $reset = () => {
    startDate.value = null;
    weeks.value = [];
    challenges.value = [];
  };

  return {
    achievements,
    allVeggies,
    atMostVeggies,
    challenges,
    completedChallenges,
    currentChallenge,
    currentVeggies,
    favorites,
    getWeekStarts,
    hotStreak,
    over30Veggies,
    startDate,
    suggestions,
    uniqueVeggies,
    veggiesForWeek,
    weeklyAchievements,
    weeks,
    setVeggiesForWeek,
    toggleVeggie,
    toggleVeggieForWeek,
    $reset,
  };
});
