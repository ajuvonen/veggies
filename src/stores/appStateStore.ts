import {ref, watchEffect} from 'vue';
import {defineStore, storeToRefs} from 'pinia';
import {useStorage} from '@vueuse/core';
import type {Settings} from '@/utils/types';
import {useAchievements} from '@/hooks/achievements';
import {useActivityStore} from '@/stores/activityStore';

type Message = {
  id: string;
  text: string;
};

export const useAppStateStore = defineStore('appState', () => {
  const {advance, achievements, reset} = useAchievements();
  const {uniqueVeggies} = storeToRefs(useActivityStore());

  watchEffect(() => {
    advance(uniqueVeggies.value, 0);
  });

  // State refs
  const settings = useStorage<Settings>(
    'veggies-settings',
    {
      locale: 'en',
    },
    localStorage,
    {
      mergeDefaults: true,
    },
  );

  const messages = ref<Message[]>([]);

  const addToastMessage = (text: string) => {
    messages.value.push({
      id: crypto.randomUUID(),
      text,
    });
  };

  const removeToastMessage = (targetId: string) => {
    messages.value = messages.value.filter(({id}) => id !== targetId);
  };

  const $reset = () => {
    messages.value = [];
    settings.value = {
      locale: 'en',
    };
    reset();
  };

  return {
    messages,
    settings,
    achievements,
    addToastMessage,
    removeToastMessage,
    $reset,
  };
});
