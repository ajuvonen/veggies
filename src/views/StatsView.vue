<script lang="ts" setup>
import {ref} from 'vue';
import {storeToRefs} from 'pinia';
import {TabGroup, TabList, Tab, TabPanels, TabPanel} from '@headlessui/vue';
import {hideAll} from 'vue-tippy';
import {useThrottleFn} from '@vueuse/core';
import {useActivityStore} from '@/stores/activityStore';
import type {IconString} from '@/components/IconComponent.vue';
import WeeklyAmountsChart from '@/components/charts/WeeklyAmountsChart.vue';
import WeeklyCategoriesChart from '@/components/charts/WeeklyCategoriesChart.vue';
import AllTimeStatus from '@/components/AllTimeStatus.vue';
import CategoryStatusChart from '@/components/charts/CategoryStatusChart.vue';
import AchievementList from '@/components/AchievementList.vue';
import VeggieList from '@/components/VeggieList.vue';
import WeekEditor from '@/components/WeekEditor.vue';

const {allVeggies, uniqueVeggies, favorites, achievements} = storeToRefs(useActivityStore());

const selectedStat = ref(0);

const icons: IconString[] = [
  'calendarWeekOutline',
  'chartLine',
  'history',
  'formatListChecks',
  'trophyOutline',
] as const;

const hideTooltips = useThrottleFn(hideAll);
</script>
<template>
  <TabGroup :selectedIndex="selectedStat" @change="(index) => (selectedStat = index)">
    <TabList class="grid grid-cols-5 gap-2">
      <Tab v-for="(_, index) in [...Array(5)]" :key="index" v-slot="{selected}" as="template">
        <ButtonComponent
          v-tippy="$t(`stats.${index}`)"
          :aria-label="$t(`stats.${index}`)"
          :class="{'!bg-[--color-highlight-dark]': selected}"
          :data-test-id="`stats-tab-${index}`"
          class="justify-center"
        >
          <IconComponent :icon="icons[index]" />
        </ButtonComponent>
      </Tab>
    </TabList>
    <TabPanels class="flex grow min-h-0">
      <TabPanel :as="WeekEditor" class="stats__tab" @scroll="hideTooltips" />
      <TabPanel class="stats__tab">
        <WeeklyAmountsChart />
        <WeeklyCategoriesChart />
      </TabPanel>
      <TabPanel class="stats__tab">
        <AllTimeStatus />
        <CategoryStatusChart totals :favorites="favorites" :veggies="allVeggies" />
      </TabPanel>
      <TabPanel
        :as="VeggieList"
        :uniqueVeggies="uniqueVeggies"
        class="stats__tab"
        @scroll="hideTooltips"
      />
      <TabPanel
        :as="AchievementList"
        :achievements="achievements"
        class="stats__tab"
        @scroll="hideTooltips"
      />
    </TabPanels>
  </TabGroup>
</template>
<style scoped>
.stats__tab {
  @apply w-full;
  @apply flex-container flex-col gap-4;
}
</style>
