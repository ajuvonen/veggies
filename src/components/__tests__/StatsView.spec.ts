import {describe, it, expect, beforeEach} from 'vitest';
import {mount} from '@vue/test-utils';
import {DateTime} from 'luxon';
import {useActivityStore} from '@/stores/activityStore';
import StatsView from '@/views/StatsView.vue';

describe('StatsView', () => {
  let activityStore: ReturnType<typeof useActivityStore>;

  beforeEach(() => {
    activityStore = useActivityStore();
  });

  it('renders current week by default', () => {
    activityStore.startDate = DateTime.now().startOf('week');
    activityStore.weeks.push({
      startDate: DateTime.now().startOf('week'),
      veggies: ['apple', 'raspberry', 'chickpea'],
    });
    const wrapper = mount(StatsView);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders five previous weeks', async () => {
    activityStore.startDate = DateTime.now().startOf('week');
    activityStore.weeks.push({
      startDate: DateTime.now().startOf('week'),
      veggies: ['apple', 'raspberry', 'chickpea'],
    });
    const wrapper = mount(StatsView);

    await wrapper.findByTestId('stats-dropdown-button').trigger('click');
    await wrapper.findByTestId('stats-dropdown-option-1').trigger('click');
    expect(wrapper.find('#weekly-amounts-table').exists()).toBe(true);
    expect(wrapper.find('#weekly-categories-table').exists()).toBe(true);
  });

  it('renders all time data', async () => {
    activityStore.startDate = DateTime.now().startOf('week');
    activityStore.weeks.push({
      startDate: DateTime.now().startOf('week'),
      veggies: ['apple', 'raspberry', 'chickpea'],
    });
    const wrapper = mount(StatsView);

    await wrapper.findByTestId('stats-dropdown-button').trigger('click');
    await wrapper.findByTestId('stats-dropdown-option-2').trigger('click');
    expect(wrapper.find('#category-status-table').exists()).toBe(true);
  });

  it('renders achievements', async () => {
    activityStore.startDate = DateTime.now().startOf('week');
    activityStore.weeks.push({
      startDate: DateTime.now().startOf('week'),
      veggies: ['apple', 'raspberry', 'chickpea'],
    });
    const wrapper = mount(StatsView);

    await wrapper.findByTestId('stats-dropdown-button').trigger('click');
    await wrapper.findByTestId('stats-dropdown-option-3').trigger('click');
    expect(wrapper.findByTestId('achievement-list').exists()).toBe(true);
  });

  it('renders veggie list', async () => {
    activityStore.startDate = DateTime.now().startOf('week');
    activityStore.weeks.push({
      startDate: DateTime.now().startOf('week'),
      veggies: ['apple', 'raspberry', 'chickpea'],
    });
    const wrapper = mount(StatsView);

    await wrapper.findByTestId('stats-dropdown-button').trigger('click');
    await wrapper.findByTestId('stats-dropdown-option-4').trigger('click');
    expect(wrapper.findByTestId('veggie-list').exists()).toBe(true);
  });
});
