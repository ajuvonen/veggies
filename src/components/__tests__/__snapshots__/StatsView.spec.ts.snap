// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatsView > renders current week by default 1`] = `
"<h1 data-v-a19ac338="" class="sr-only">Statistics</h1>
<button aria-hidden="true" style="position: fixed; height: 0px; padding: 0px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;" type="button"></button>
<div data-v-a19ac338="" role="tablist" aria-orientation="horizontal" class="grid grid-cols-5 gap-2"><button data-v-24e38f3d="" data-v-a19ac338="" class="button button--primary !bg-sky-700 justify-center" aria-label="This week's veggies" data-test-id="stats-tab-0" id="headlessui-tabs-tab-v-0" role="tab" aria-selected="true" tabindex="0" data-headlessui-state="selected">
    <!--v-if--><svg data-v-a19ac338="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
      <path d="M5 3H6V1H8V3H16V1H18V3H19C20.11 3 21 3.89 21 5V19C21 20.1 20.11 21 19 21H5C3.9 21 3 20.11 3 19V5C3 3.89 3.9 3 5 3M5 19H19V9H5V19M5 7H19V5H5V7M17 11V13H7V11H17" class="transition-all"></path>
    </svg>
  </button><button data-v-24e38f3d="" data-v-a19ac338="" class="button button--primary justify-center" aria-label="Weekly statistics" data-test-id="stats-tab-1" id="headlessui-tabs-tab-v-1" role="tab" aria-selected="false" tabindex="-1" data-headlessui-state="">
    <!--v-if--><svg data-v-a19ac338="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
      <path d="M16,11.78L20.24,4.45L21.97,5.45L16.74,14.5L10.23,10.75L5.46,19H22V21H2V3H4V17.54L9.5,8L16,11.78Z" class="transition-all"></path>
    </svg>
  </button><button data-v-24e38f3d="" data-v-a19ac338="" class="button button--primary justify-center" aria-label="All time statistics" data-test-id="stats-tab-2" id="headlessui-tabs-tab-v-2" role="tab" aria-selected="false" tabindex="-1" data-headlessui-state="">
    <!--v-if--><svg data-v-a19ac338="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
      <path d="M13.5,8H12V13L16.28,15.54L17,14.33L13.5,12.25V8M13,3A9,9 0 0,0 4,12H1L4.96,16.03L9,12H6A7,7 0 0,1 13,5A7,7 0 0,1 20,12A7,7 0 0,1 13,19C11.07,19 9.32,18.21 8.06,16.94L6.64,18.36C8.27,20 10.5,21 13,21A9,9 0 0,0 22,12A9,9 0 0,0 13,3" class="transition-all"></path>
    </svg>
  </button><button data-v-24e38f3d="" data-v-a19ac338="" class="button button--primary justify-center" aria-label="Veggie list" data-test-id="stats-tab-3" id="headlessui-tabs-tab-v-3" role="tab" aria-selected="false" tabindex="-1" data-headlessui-state="">
    <!--v-if--><svg data-v-a19ac338="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
      <path d="M3,5H9V11H3V5M5,7V9H7V7H5M11,7H21V9H11V7M11,15H21V17H11V15M5,20L1.5,16.5L2.91,15.09L5,17.17L9.59,12.59L11,14L5,20Z" class="transition-all"></path>
    </svg>
  </button><button data-v-24e38f3d="" data-v-a19ac338="" class="button button--primary justify-center" aria-label="Achievements" data-test-id="stats-tab-4" id="headlessui-tabs-tab-v-4" role="tab" aria-selected="false" tabindex="-1" data-headlessui-state="">
    <!--v-if--><svg data-v-a19ac338="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
      <path d="M18 2C17.1 2 16 3 16 4H8C8 3 6.9 2 6 2H2V11C2 12 3 13 4 13H6.2C6.6 15 7.9 16.7 11 17V19.08C8 19.54 8 22 8 22H16C16 22 16 19.54 13 19.08V17C16.1 16.7 17.4 15 17.8 13H20C21 13 22 12 22 11V2H18M6 11H4V4H6V11M16 11.5C16 13.43 15.42 15 12 15C8.59 15 8 13.43 8 11.5V6H16V11.5M20 11H18V4H20V11Z" class="transition-all"></path>
    </svg>
  </button></div>
<div data-v-a19ac338="" class="flex flex-grow min-h-0">
  <div data-v-81c33b70="" data-v-a19ac338="" class="week-editor stats__tab" id="headlessui-tabs-panel-v-5" role="tabpanel" tabindex="0" data-headlessui-state="selected">
    <div class="relative z-30" data-headlessui-state="">
      <div data-v-81c33b70="" class="flex-container flex-col"><label id="headlessui-listbox-label-v-6" data-headlessui-state="" class="label-like">Choose Week</label>
        <listbox-button-stub data-v-81c33b70="" as="button" class="week-editor__dropdown-button" data-test-id="week-editor-dropdown-button"></listbox-button-stub>
      </div>
      <!---->
    </div>
    <!--v-if-->
    <transition-group-stub data-v-cd627eac="" data-v-81c33b70="" name="tags" tag="ul" appear="false" persisted="false" css="true" class="tags__container">
      <li data-v-cd627eac="" data-test-id="tag-apple" class="z-10"><button data-v-24e38f3d="" data-v-cd627eac="" class="button button--tag button--remove" aria-label="Remove apple">
          <!--v-if--><svg data-v-cd627eac="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
            <path d="M19,13H5V11H19V13Z" class="transition-all"></path>
          </svg> apple
        </button></li>
      <li data-v-cd627eac="" data-test-id="tag-chickpea" class="z-10"><button data-v-24e38f3d="" data-v-cd627eac="" class="button button--tag button--remove" aria-label="Remove chickpea">
          <!--v-if--><svg data-v-cd627eac="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
            <path d="M19,13H5V11H19V13Z" class="transition-all"></path>
          </svg> chickpea
        </button></li>
      <li data-v-cd627eac="" data-test-id="tag-raspberry" class="z-10"><button data-v-24e38f3d="" data-v-cd627eac="" class="button button--tag button--remove" aria-label="Remove raspberry">
          <!--v-if--><svg data-v-cd627eac="" style="width: 1.25rem; height: 1.25rem;" class="flex-shrink-0" viewBox="0 0 24 24" aria-hidden="true">
            <path d="M19,13H5V11H19V13Z" class="transition-all"></path>
          </svg> raspberry
        </button></li>
    </transition-group-stub>
  </div><span data-v-a19ac338="" aria-hidden="true" style="position: fixed; height: 0px; padding: 0px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;" id="headlessui-tabs-panel-v-7" role="tabpanel" tabindex="-1" class="stats__tab"></span><span data-v-a19ac338="" aria-hidden="true" style="position: fixed; height: 0px; padding: 0px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;" id="headlessui-tabs-panel-v-8" role="tabpanel" tabindex="-1" class="stats__tab"></span><span data-v-a19ac338="" aria-hidden="true" style="position: fixed; height: 0px; padding: 0px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;" id="headlessui-tabs-panel-v-9" role="tabpanel" tabindex="-1" uniqueveggies="apple,raspberry,chickpea" class="stats__tab"></span><span data-v-a19ac338="" aria-hidden="true" style="position: fixed; height: 0px; padding: 0px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;" id="headlessui-tabs-panel-v-10" role="tabpanel" tabindex="-1" class="stats__tab"></span>
</div>"
`;
